synchronise fileBase subclasses use of position

grandinj opened the pull request:
needs to be threadsafe, so we need to lock it.
Put the relevant code into a new class FileBaseDefault

grandinj committed:
synchronise fileBase subclasses use of position

needs to be threadsafe, so we need to lock it.
Put the relevant code into a new class FileBaseDefault

andreitokar commented on h2/src/main/org/h2/store/fs/FileBaseDefault.java:
```diff
@@ -0,0 +1,58 @@
+    
+    @Override
+    public final synchronized int read(ByteBuffer dst) throws IOException {
+        int read = read(dst, position);
```
Why that read() has to be under lock? We only need to protect usage of the "position" field (or even make it atomic instead) and let actual read() / write() concurrency to be taken care of by underlying FileChannel.

grandinj replied:
(1) this read/write is only rarely used, so is not performance critical
(2) depending on the specific subclass, may not have something underneath it that is synchronised.
(3) we don't know exactly how much will actually be read/written, since the subclass may read/write less than requested, so we need to lock over the whole operation.

andreitokar replied:
Sorry, I missed that item (3). Thanks!

andreitokar commented on h2/src/main/org/h2/store/fs/FileBaseDefault.java:
```diff
@@ -0,0 +1,58 @@
+    
+    @Override
+    public synchronized final long position() throws IOException {
+        return 0;
```
"return position;" maybe?

andreitokar commented on h2/src/main/org/h2/store/fs/retry/FileRetryOnInterrupt.java:
```diff
@@ -82,17 +82,6 @@ public void implCloseChannel() throws IOException {
-            try {
-                return channel.position();
-            } catch (IOException e) {
-                reopen(i, e);
```
Not sure, if it is possible at all to be interrupted and get ClosedChannelException during this operation,  but that logic is lost now, isn't it?

andreitokar commented on h2/src/main/org/h2/store/fs/retry/FileRetryOnInterrupt.java:
```diff
@@ -129,23 +105,11 @@ public int read(ByteBuffer dst, long position) throws IOException {
-                channel.position(pos);
-                return this;
-            } catch (IOException e) {
-                reopen(i, e);
```
Same thing here, no retry anymore.

grandinj committed:
restore retry logic

grandinj committed:
no need to extend FileBaseDefault here

the underlying channel already locks position, no need to store that field twice

grandinj committed:
this should let NonWritableChannelException through

to be consistent with how java's FileChannel works, we don't have a test for this, which is why it snuck through, but was revealed now that we are sharing this method with the normal read(ByteBuffer)

grandinj committed:
no need to use FileBaseDefault here

underlying FileChannel already locks and stores position

grandinj committed:
fix position method, check returns properly

andreitokar reviewed (approved):
LGTM
